## Objective: The objective of this assignment is to assess students' understanding of batch normalization in artificial neural networks (ANN) and its impact on training performance.

In [ ]:
Q1. Theory and Concepts
1. Explain the concept of batch normalization in the context of Artificial Neural Networks
2. Describe the benefits of using batch normalization during training
3. Discuss the working principle of batch normalization, including the normalization step and the learnable 
parameters.

1. Batch Norm is a normalization technique done between the layers of a Neural Network instead of in the raw data. It is done along mini-batches instead of the full data set. It serves to speed up training and use higher learning rates, making learning easier.
Batch-Normalization (BN) is an algorithmic method which makes the training of Deep Neural Networks (DNN) faster and more stable

2. Using batch normalization allows us to use much higher learning rates, which further increases the speed at which networks train. Makes weights easier to initialize — Weight initialization can be difficult, and it's even more difficult when creating deeper networks

3. It consists of normalizing activation vectors from hidden layers using the first and the second statistical moments (mean and variance) of the current batch. This normalization step is applied right before (or right after) the nonlinear function.


Batch Norm is just another network layer that gets inserted between a hidden layer and the next hidden layer. Its job is to take the outputs from the first hidden layer and normalize them before passing them on as the input of the next hidden layer.

Just like the parameters (eg. weights, bias) of any network layer, a Batch Norm layer also has parameters of its own:
    
Two learnable parameters called beta and gamma.
Two non-learnable parameters (Mean Moving Average and Variance Moving Average) are saved as part of the ‘state’ of the Batch Norm layer.


These parameters are per Batch Norm layer. So if we have, say, three hidden layers and three Batch Norm layers in the network, we would have three learnable beta and gamma parameters for the three layers. Similarly for the Moving Average parameters.

During training, we feed the network one mini-batch of data at a time. During the forward pass, each layer of the network processes that mini-batch of data. The Batch Norm layer processes its data as follows.

In [ ]:
Q2. Implementation
1. Choose a dataset of your choice (e.g., MNIST, CIFAR-10) and preprocess it
2. Implement a simple feedforward neural network using any deep learning framework/library (e.g., 
TensorFlow, PyTorch)
3. Train the neural network on the chosen dataset without using batch normalization
4. Implement batch normalization layers in the neural network and train the model again
5. Compare the training and validation performance (e.g., accuracy, loss) between the models with and 
without batch normalizationr
6 Discuss the impact of batch normalization on the training process and the performance of the neural 
network.


## With batch normalization (size 32)

In [7]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

mnist=tf.keras.datasets.mnist
(X_train_full, y_train_full), (X_test, y_test) = mnist.load_data()

X_valid,X_train=X_train_full[:5000]/255., X_train_full[5000:]/255.
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

# scale the test set as well
X_test = X_test / 255.

# Creating layers of ANN
LAYERS=[tf.keras.layers.Flatten(input_shape=[28, 28], name="inputLayer"),
        tf.keras.layers.Dense(300, activation="relu", name="hiddenLayer1"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100, activation="relu", name="hiddenLayer2"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10, activation="softmax", name="outputLayer")]

model_clf=tf.keras.models.Sequential(LAYERS)

LOSS_FUNCTION = "sparse_categorical_crossentropy" # use => tf.losses.sparse_categorical_crossentropy
OPTIMIZER = "SGD" # or use with custom learning rate=> tf.keras.optimizers.SGD(0.02)
METRICS = ["accuracy"]


model_clf.compile(loss=LOSS_FUNCTION,
              optimizer=OPTIMIZER,
              metrics=METRICS)



EPOCHS = 10
VALIDATION_SET = (X_valid, y_valid)

history=model_clf.fit(X_train,y_train,epochs=EPOCHS,validation_data=VALIDATION_SET,batch_size=32)

pd.DataFrame(history.history)

Epoch 1/10
1719/1719 [==============================] - 8s 4ms/step - loss: 0.3395 - accuracy: 0.9012 - val_loss: 0.1523 - val_accuracy: 0.9570
Epoch 2/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.1688 - accuracy: 0.9514 - val_loss: 0.1128 - val_accuracy: 0.9658
Epoch 3/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.1247 - accuracy: 0.9639 - val_loss: 0.0992 - val_accuracy: 0.9704
Epoch 4/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.0987 - accuracy: 0.9714 - val_loss: 0.0923 - val_accuracy: 0.9702
Epoch 5/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.0821 - accuracy: 0.9761 - val_loss: 0.0851 - val_accuracy: 0.9740
Epoch 6/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.0671 - accuracy: 0.9806 - val_loss: 0.0818 - val_accuracy: 0.9738
Epoch 7/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.0582 - accuracy: 0.9832 - val_loss: 0.0816 - val_accuracy:

loss  accuracy  val_loss  val_accuracy
0  0.339461  0.901236  0.152341        0.9570
1  0.168787  0.951400  0.112844        0.9658
2  0.124734  0.963855  0.099177        0.9704
3  0.098705  0.971364  0.092308        0.9702
4  0.082098  0.976055  0.085137        0.9740
5  0.067097  0.980600  0.081810        0.9738
6  0.058196  0.983182  0.081604        0.9768
7  0.049940  0.985727  0.077088        0.9768
8  0.042007  0.988509  0.078856        0.9750
9  0.038037  0.989345  0.077891        0.9772

## Without batch normalization

import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

mnist=tf.keras.datasets.mnist
(X_train_full, y_train_full), (X_test, y_test) = mnist.load_data()

X_valid,X_train=X_train_full[:5000]/255., X_train_full[5000:]/255.
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

# scale the test set as well
X_test = X_test / 255.

# Creating layers of ANN
LAYERS=[tf.keras.layers.Flatten(input_shape=[28, 28], name="inputLayer"),
        tf.keras.layers.Dense(300, activation="relu", name="hiddenLayer1"),
        tf.keras.layers.Dense(100, activation="relu", name="hiddenLayer2"),
        tf.keras.layers.Dense(10, activation="softmax", name="outputLayer")]

model_clf=tf.keras.models.Sequential(LAYERS)

LOSS_FUNCTION = "sparse_categorical_crossentropy" # use => tf.losses.sparse_categorical_crossentropy
OPTIMIZER = "SGD" # or use with custom learning rate=> tf.keras.optimizers.SGD(0.02)
METRICS = ["accuracy"]


model_clf.compile(loss=LOSS_FUNCTION,
              optimizer=OPTIMIZER,
              metrics=METRICS)



EPOCHS = 10
VALIDATION_SET = (X_valid, y_valid)

history=model_clf.fit(X_train,y_train,epochs=EPOCHS,validation_data=VALIDATION_SET,batch_size=32)

pd.DataFrame(history.history)

## accuracy is increased with batch normalization from 0.969 to 0.989 (for training data) and 0.967 to 0.977 (for validation data)

## loss is reduced with batch normalization from 0.108 to 0.038 (for training data) and 0.111 to 0.077 (for validation data)

In [ ]:
Q3. Experimentation and Analysis
1. Experiment with different batch sizes and observe the effect on the training dynamics and model 
performancer
2. Discuss the advantages and potential limitations of batch normalization in improving the training of 
neural networks.

## With batch normalization (size 64)

In [9]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

mnist=tf.keras.datasets.mnist
(X_train_full, y_train_full), (X_test, y_test) = mnist.load_data()

X_valid,X_train=X_train_full[:5000]/255., X_train_full[5000:]/255.
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

# scale the test set as well
X_test = X_test / 255.

# Creating layers of ANN
LAYERS=[tf.keras.layers.Flatten(input_shape=[28, 28], name="inputLayer"),
        tf.keras.layers.Dense(300, activation="relu", name="hiddenLayer1"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100, activation="relu", name="hiddenLayer2"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10, activation="softmax", name="outputLayer")]

model_clf=tf.keras.models.Sequential(LAYERS)

LOSS_FUNCTION = "sparse_categorical_crossentropy" # use => tf.losses.sparse_categorical_crossentropy
OPTIMIZER = "SGD" # or use with custom learning rate=> tf.keras.optimizers.SGD(0.02)
METRICS = ["accuracy"]


model_clf.compile(loss=LOSS_FUNCTION,
              optimizer=OPTIMIZER,
              metrics=METRICS)



EPOCHS = 10
VALIDATION_SET = (X_valid, y_valid)

history=model_clf.fit(X_train,y_train,epochs=EPOCHS,validation_data=VALIDATION_SET,batch_size=64)

pd.DataFrame(history.history)

Epoch 1/10
860/860 [==============================] - 6s 5ms/step - loss: 0.4000 - accuracy: 0.8829 - val_loss: 0.2059 - val_accuracy: 0.9436
Epoch 2/10
860/860 [==============================] - 3s 4ms/step - loss: 0.1941 - accuracy: 0.9463 - val_loss: 0.1555 - val_accuracy: 0.9548
Epoch 3/10
860/860 [==============================] - 3s 3ms/step - loss: 0.1476 - accuracy: 0.9578 - val_loss: 0.1302 - val_accuracy: 0.9616
Epoch 4/10
860/860 [==============================] - 3s 3ms/step - loss: 0.1184 - accuracy: 0.9664 - val_loss: 0.1167 - val_accuracy: 0.9644
Epoch 5/10
860/860 [==============================] - 3s 3ms/step - loss: 0.0999 - accuracy: 0.9719 - val_loss: 0.1064 - val_accuracy: 0.9680
Epoch 6/10
860/860 [==============================] - 3s 3ms/step - loss: 0.0847 - accuracy: 0.9765 - val_loss: 0.0994 - val_accuracy: 0.9686
Epoch 7/10
860/860 [==============================] - 3s 3ms/step - loss: 0.0732 - accuracy: 0.9802 - val_loss: 0.0945 - val_accuracy: 0.9710
Epoch 

loss  accuracy  val_loss  val_accuracy
0  0.400043  0.882927  0.205926        0.9436
1  0.194065  0.946345  0.155528        0.9548
2  0.147613  0.957800  0.130193        0.9616
3  0.118389  0.966382  0.116655        0.9644
4  0.099855  0.971891  0.106387        0.9680
5  0.084659  0.976491  0.099388        0.9686
6  0.073190  0.980182  0.094548        0.9710
7  0.064344  0.983345  0.091589        0.9710
8  0.056262  0.985327  0.087270        0.9734
9  0.049713  0.986945  0.086281        0.9740

## With batch normalization (size 16)

In [10]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

mnist=tf.keras.datasets.mnist
(X_train_full, y_train_full), (X_test, y_test) = mnist.load_data()

X_valid,X_train=X_train_full[:5000]/255., X_train_full[5000:]/255.
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

# scale the test set as well
X_test = X_test / 255.

# Creating layers of ANN
LAYERS=[tf.keras.layers.Flatten(input_shape=[28, 28], name="inputLayer"),
        tf.keras.layers.Dense(300, activation="relu", name="hiddenLayer1"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100, activation="relu", name="hiddenLayer2"),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(10, activation="softmax", name="outputLayer")]

model_clf=tf.keras.models.Sequential(LAYERS)

LOSS_FUNCTION = "sparse_categorical_crossentropy" # use => tf.losses.sparse_categorical_crossentropy
OPTIMIZER = "SGD" # or use with custom learning rate=> tf.keras.optimizers.SGD(0.02)
METRICS = ["accuracy"]


model_clf.compile(loss=LOSS_FUNCTION,
              optimizer=OPTIMIZER,
              metrics=METRICS)



EPOCHS = 10
VALIDATION_SET = (X_valid, y_valid)

history=model_clf.fit(X_train,y_train,epochs=EPOCHS,validation_data=VALIDATION_SET,batch_size=16)

pd.DataFrame(history.history)

Epoch 1/10
3438/3438 [==============================] - 13s 3ms/step - loss: 0.3112 - accuracy: 0.9062 - val_loss: 0.1335 - val_accuracy: 0.9640
Epoch 2/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.1633 - accuracy: 0.9514 - val_loss: 0.1062 - val_accuracy: 0.9696
Epoch 3/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.1246 - accuracy: 0.9627 - val_loss: 0.0858 - val_accuracy: 0.9750
Epoch 4/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.1008 - accuracy: 0.9690 - val_loss: 0.0774 - val_accuracy: 0.9780
Epoch 5/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.0860 - accuracy: 0.9740 - val_loss: 0.0746 - val_accuracy: 0.9794
Epoch 6/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.0763 - accuracy: 0.9766 - val_loss: 0.0696 - val_accuracy: 0.9814
Epoch 7/10
3438/3438 [==============================] - 9s 3ms/step - loss: 0.0663 - accuracy: 0.9795 - val_loss: 0.0640 - val_accuracy

loss  accuracy  val_loss  val_accuracy
0  0.311245  0.906182  0.133507        0.9640
1  0.163277  0.951382  0.106187        0.9696
2  0.124562  0.962727  0.085798        0.9750
3  0.100842  0.969000  0.077413        0.9780
4  0.086010  0.973964  0.074630        0.9794
5  0.076302  0.976636  0.069615        0.9814
6  0.066287  0.979545  0.064046        0.9812
7  0.060520  0.981127  0.066720        0.9820
8  0.051045  0.983764  0.069149        0.9816
9  0.047922  0.984927  0.072400        0.9794

## Conclusion:
    1. batch size 32 get better results than batch size 16 and 32, 
    2. so we need to use batch size as hyper parameter to get the optimum batch size.